packages/vite/src/node/optimizer/esbuildDepPlugin.ts

```diff
@@ -209,10 +213,18 @@ export function esbuildDepPlugin(
       if (isRunningWithYarnPnp) {
         build.onResolve(
           { filter: /.*/ },
-          async ({ path, importer, kind, resolveDir }) => ({
-            // pass along resolveDir for entries
-            path: await resolve(path, importer, kind, resolveDir)
-          })
+          async ({ path: id, importer, kind, resolveDir, namespace }) => {
+            const resolved = await resolve(
+              id,
+              importer,
+              kind,
+              // pass along resolveDir for entries
+              namespace === 'dep' ? resolveDir : undefined
+            )
+            if (resolved) {
+              return resolveResult(id, resolved)
+            }
+          }
```

Comment:
Added handling of browser-external and external URLs here as well.

Without this packages with browser mappings to `false` fail to pre-bundle.

However, it doesn't completely solves the problem, as can be seen with `@loaders.gl/loader-utils`:

```
✘ [ERROR] No matching export in "browser-external:./lib/node/util" for import "promisify"

    .yarn/cache/@loaders.gl-loader-utils-npm-3.1.6-047c300f44-ba6ead7208.zip/node_modules/@loaders.gl/loader-utils/dist/esm/index.js:22:9:
      22 │ export { promisify } from './lib/node/util';
         ╵          ~~~~~~~~~
```

According to this issue (https://github.com/vitejs/vite/issues/2320) it's expected behavior. But it _does work_ with PnP disabled. 

Here's repro:
https://github.com/swandir/vite-browser-external-pnp

Seems like ESBuild itself successfully handles with case when dealing with file system directly, but fails when going through `onResolve`.

I haven't figured out yet what is the difference here exactly.

Reply:
Code that caused this failure:

https://github.com/visgl/loaders.gl/blob/4fe0724b942697a95ff3eb3af47b7a7c18556c52/modules/loader-utils/src/index.ts#L87-L95

These `./lib/node/*` modules are mapped to `false` via `browser` field, and it works fine with star import. But breaks when importing code references named exports.